grammar.js

```diff
@@ -1301,6 +1301,71 @@ module.exports = grammar({
       alias($._literal_string, $.literal),
     ),
 
+    _operator_class: $ => seq(
+      field("opclass", $.identifier),
+      optional(
+        field("opclass_parameters", wrapped_in_parenthesis(comma_list($.term)))
+      )
+    ),
+
+    _index_column: $ => seq(
+      field("name", $._column),
+      optional(seq($.keyword_collate, $.identifier)),
+      optional($._operator_class),
+      optional($.direction),
+      optional(
+        seq(
+          $.keyword_nulls,
+          choice(
+            $.keyword_first,
+            $.keyword_last
+          )
+        )
+      )
+    ),
+
+    _index_expression: $ => seq(
+      wrapped_in_parenthesis($._expression),
+      optional(seq($.keyword_collate, $.identifier)),
+      optional($._operator_class),
+      optional($.direction),
+      optional(
+        seq(
+          $.keyword_nulls,
+          choice(
+            $.keyword_first,
+            $.keyword_last
+          )
+        )
+      )
+    ),
+
+    _index_column_function: $ => seq(
+      $.invocation,
+      optional(seq($.keyword_collate, $.identifier)),
+      optional($._operator_class),
+      optional($.direction),
+      optional(
+        seq(
+          $.keyword_nulls,
+          choice(
+            $.keyword_first,
+            $.keyword_last
+          )
+        )
+      )
+    ),
+
+    _index_field: $ => seq(
+      choice(
+        alias($._index_column, $.column),
+        alias($._index_expression, $.column_expression),
+        alias($._index_column_function, $.column_function)
```

Comment:
I like your suggestion of moving the duplication into this `_index_field` node. I'd suggest using a `field` for each type of field like this:

```suggestion
        field("expression", wrapped_in_parenthesis($._expression)),
        field("function", $.invocation),
        field("column", $._column),
```